state/protocol/badger/mutator.go

```diff
@@ -85,32 +88,52 @@ func NewFullConsensusState(
 	blockTimer protocol.BlockTimer,
 	receiptValidator module.ReceiptValidator,
 	sealValidator module.SealValidator,
-) (*MutableState, error) {
+) (*ParticipantState, error) {
 	followerState, err := NewFollowerState(state, index, payloads, tracer, consumer, blockTimer)
 	if err != nil {
 		return nil, fmt.Errorf("initialization of Mutable Follower State failed: %w", err)
 	}
-	return &MutableState{
+	return &ParticipantState{
 		FollowerState:    followerState,
 		receiptValidator: receiptValidator,
 		sealValidator:    sealValidator,
 	}, nil
 }
 
-// Extend extends the protocol state of a CONSENSUS FOLLOWER. While it checks
+// ExtendCertified extends the protocol state of a CONSENSUS FOLLOWER. While it checks
 // the validity of the header; it does _not_ check the validity of the payload.
 // Instead, the consensus follower relies on the consensus participants to
-// validate the full payload. Therefore, a follower a QC (i.e. a child block) as
-// proof that a block is valid.
-func (m *FollowerState) Extend(ctx context.Context, candidate *flow.Block) error {
-
+// validate the full payload. Payload validity can be proved by a valid quorum certificate.
+// Certifying QC must match candidate block: candidate.View == certifyingQC.View && candidate.ID() == certifyingQC.BlockID
+// NOTE: this function expects that `certifyingQC` has been validated.
+// Expected errors during normal operations:
+//   - state.OutdatedExtensionError if the candidate block is outdated (e.g. orphaned)
+func (m *FollowerState) ExtendCertified(ctx context.Context, candidate *flow.Block, certifyingQC *flow.QuorumCertificate) error {
 	span, ctx := m.tracer.StartSpanFromContext(ctx, trace.ProtoStateMutatorHeaderExtend)
 	defer span.End()
 
+	// TODO: this is a temporary if statement since follower engine doesn't deliver QCs yet. Once the implementation is complete
+	// there are no cases where certifyingQC can be nil.
+	if certifyingQC != nil {
+		blockID := candidate.ID()
+		// sanity check if certifyingQC actually certifies candidate block
+		if certifyingQC.View != candidate.Header.View {
+			return fmt.Errorf("qc doesn't certify candidate block, expect %d view, got %d", candidate.Header.View, certifyingQC.View)
+		}
+		if certifyingQC.BlockID != blockID {
+			return fmt.Errorf("qc doesn't certify candidate block, expect %x blockID, got %x", blockID, certifyingQC.BlockID)
+		}
+	}
+
 	// check if the block header is a valid extension of the finalized state
 	err := m.headerExtend(candidate)
 	if err != nil {
-		return fmt.Errorf("header not compliant with chain state: %w", err)
+		if state.IsOutdatedExtensionError(err) {
+			return fmt.Errorf("candidate block is an outdated extension: %w", err)
+		}
+		// since we have a QC for this block, it cannot be an invalid extension
+		return fmt.Errorf("unexpected invalid block (id=%x) with certifying qc (id=%x): %s",
+			candidate.ID(), certifyingQC.ID(), err.Error())
```

Comment:
As follow-up on my comment I think you can just keep it short
```suggestion
		// In accordance with API specification, we only expect `OutdatedExtensionError` here:
		return fmt.Errorf("header not compliant with chain state: %w", err)
```